src/main/scala/beam/analysis/cartraveltime/CarRideStatsFromPathTraversalEventHandler.scala

```diff
@@ -0,0 +1,301 @@
+    IterationCarRideStats(
+      iteration = iterationNumber,
+      travelTime = travelTimeStas,
+      speed = speedStats,
+      distance = distanceStats,
+      freeFlowTravelTime = freeFlowTravelTimeStats,
+      freeFlowSpeed = freeFlowSpeedStats
+    )
+  }
+
+  def getRideStats(
+    networkHelper: NetworkHelper,
+    freeFlowTravelTime: FreeFlowTravelTime,
+    iterationNumber: Int,
+    carPtes: Seq[PathTraversalEvent]
+  ): Seq[SingleRideStat] = {
+    val drivingWithParkingPtes = buildDrivingParking(carPtes)
+    val stats = buildRideStats(networkHelper, freeFlowTravelTime, drivingWithParkingPtes)
+    logger.info(
+      s"For the iteration ${iterationNumber} created ${stats.length} ride stats from ${carPtes.size} PathTraversalEvents"
+    )
+    stats
+  }
+
+  def calcFreeFlowDuration(freeFlowTravelTime: FreeFlowTravelTime, linkIds: IndexedSeq[Link]): Double = {
+    linkIds.foldLeft(0.0) {
+      case (acc, link) =>
+        val t = freeFlowTravelTime.getLinkTravelTime(link, 0.0, null, null)
+        acc + t
+    }
+  }
+
+  def buildRideStats(
+    networkHelper: NetworkHelper,
+    freeFlowTravelTimeCalc: FreeFlowTravelTime,
+    drivingWithParkingPtes: Iterable[(PathTraversalEvent, PathTraversalEvent)]
+  ): Seq[SingleRideStat] = {
+    val stats = drivingWithParkingPtes.foldLeft(List.empty[SingleRideStat]) {
+      case (acc, (driving, parking)) =>
+        if (driving.arrivalTime != parking.departureTime) {
+          val msg = s"arrivalTime != departureTime\n\tdriving: $driving\n\tparking: $parking"
+          logger.warn(msg)
+        }
+        val travelTime =
+          ((driving.arrivalTime - driving.departureTime) + (parking.arrivalTime - parking.departureTime)).toDouble
+        val length = driving.legLength + parking.legLength
+        val linkIds = (driving.linkIds ++ parking.linkIds).map(lid => networkHelper.getLinkUnsafe(lid))
+        val freeFlowTravelTime: Double = calcFreeFlowDuration(freeFlowTravelTimeCalc, linkIds)
+        SingleRideStat(driving.vehicleId.toString, travelTime, length, freeFlowTravelTime) :: acc
+    }
+    stats
+  }
+
+  def buildDrivingParking(ptes: Seq[PathTraversalEvent]): Iterable[(PathTraversalEvent, PathTraversalEvent)] = {
+    val drivingWithParkingPtes = ptes
+      .groupBy(x => (x.vehicleId, x.driverId))
+      .map {
+        case ((vehId, driverId), xs) =>
+          val sorted = xs.sortBy(x => x.departureTime)
+          if (sorted.length % 2 == 1) {
+            logger.warn(s"Vehicle $vehId with driver $driverId has ${sorted.length} events")
+          }
+          sorted.sliding(2, 2).flatMap { ptes =>
+            val maybeDriving = ptes.lift(0)
+            val maybeParking = ptes.lift(1)
+            for {
+              driving <- maybeDriving
+              parking <- maybeParking
+            } yield (driving, parking)
+          }
+      }
+      .flatten
+    drivingWithParkingPtes
+  }
+
+  def main(args: Array[String]): Unit = {
```

Comment:
why the main? So it can also be run standalone for verifications?

Reply:
Yes, just for that. Easier usage in case when you already have events file from previous run